fix(dynamic-sampling): adjust rebalance interval and granularity

constantinius opened the pull request:
When sampling mode is set to `manual`, take 30d into account for rebalancing projects to have more stable sample rates


Closes: https://github.com/getsentry/projects/issues/461

constantinius committed:
fix(dynamic-sampling): adjust rebalance interval and granularity

When sampling mode is set to `manual`, take 24h into account for rebalancing projects to have more stable sample rates

jan-auer commented on src/sentry/api/endpoints/organization_details.py:
```diff
@@ -997,7 +998,20 @@ def put(self, request: Request, organization: Organization) -> Response:
                     "samplingMode" in changed_data or "targetSampleRate" in changed_data
                 ):
-                    boost_low_volume_projects_of_org_with_query.delay(organization.id)
+                    # use 1 day as the default interval for project sampling rate rebalancing
```
This code branch handles the switch to automatic mode and when the target rate is changed. See the comment above the branch. 

You'll want to modify the logic inside `_compute_project_target_sample_rates` instead.

constantinius committed:
fix(dynamic-sampling): fix initial project sampling rates calculation

Refactored project transaction count querying:
 - separation of concerns: timeout vs chunking
 - extracted storing of project rates in redis from adjusting of sample rates
 -  renamed adjust_sample_rates_of_projects to calculate_sample_rates_of_projects

Create and use type aliases for various two tuple types storing org/project/counts.

jan-auer commented on src/sentry/dynamic_sampling/tasks/boost_low_volume_projects.py:
```diff
@@ -234,92 +242,17 @@ def fetch_projects_with_total_root_transaction_count_and_rates(
+            ),
+            func_name,
+        ):
+            for row in chunk:
```
The function `query_projects_with_total_root_transaction_count_and_rates` returns a tuple without named fields. The code below must be changed to access the tuple without row field names. In the other call to this function, like so:

```py
for row in chunk:
  aggregated_projects[row[0]].append(row[1:])
```

jan-auer commented on src/sentry/dynamic_sampling/tasks/boost_low_volume_projects.py:
```diff
@@ -234,92 +242,17 @@ def fetch_projects_with_total_root_transaction_count_and_rates(
-                data = data[:-1]
-
-            for row in data:
+        for chunk in TimedIterator(
+            context,
+            query_projects_with_total_root_transaction_count_and_rates(
+                org_ids,
+                measure,
+                query_interval,
+            ),
+            func_name,
+        ):
```
_nit_: This could be more readable if written like so:

```suggestion
        iter = query_projects_with_total_root_transaction_count_and_rates(org_ids, measure, query_interval)
        for chunk in TimedIterator(context, iter, func_name):
```

constantinius committed:
fix(dynamic-sampling): work in feedback

renamed type aliases and function name for better readbility.
Improved query building: inlining several fields.
Improved readability when using TimedIterator with sub-iterator.
Fiexed field mapping for retrieved rows.

constantinius committed:
fix(dynamic-sampling): several small fixes

constantinius committed:
Merge branch 'master' into constantinius/fix/dynamic-sampling/rebalance-interval-granularity

constantinius committed:
fix(dynamic-sampling): fixing case when no rebalancing could be done

constantinius committed:
fix(dynamic-sampling): fix typing issues

jan-auer reviewed (approved):
The description reads "24h", but we're using a 30 day interval. It's best if we use the longer timeframe as implemented.

constantinius committed:
Merge branch 'master' into constantinius/fix/dynamic-sampling/rebalance-interval-granularity
